### DISCUSSION TITLE
Cache strategy SWR on Prisma Accelerate doesn't seem to work

### DISCUSSION THREAD

**Discussion by JWW87** [prisma/prisma#23787] [Q&A]
### Question

Hi all.

I have recently started a NextJS project with Prisma Accelerate. I managed to get it to work quite easily and I'm very happy with it, the performance is really good. 

However, I'm having trouble with the SWR cache strategy. The way I understand it from the docs, stale data is served from cache during the SWR lifetime but is refreshed asynchronously. I would assume that refreshed data would be served next time the request is done, but I can't get this to work.

### How to reproduce (optional)

If I run this query (simplified for readability) I get user session data and data about the accelerate cache:
```ts
const { data: sessionData, cacheInfo } = await prisma.userSession.findUnique({
  where: { id: sessionId },
  cacheStrategy: {
    ttl: 15,
    swr: 60
  }
}).withAccelerateInfo();
```
After the first request, the cacheInfo is:
```
{
  cacheStatus: 'miss',
  lastModified: 2024-04-09T12:47:09.000Z,
  region: 'AMS',
  requestId: '871a9aa11cb1b966-AMS',
  signature: '8ae5f2c7f841088afa8962daaa6578427c5045fd6839d666cbb67ba0a6cebf7b'
}
```
And after the second request it's:
```
{
  cacheStatus: 'ttl',
  lastModified: 2024-04-09T12:47:09.000Z,
  region: 'AMS',
  requestId: '871a9aca69520dfb-AMS',
  signature: '8ae5f2c7f841088afa8962daaa6578427c5045fd6839d666cbb67ba0a6cebf7b'
}
```
After 15 seconds:
```
{
  cacheStatus: 'swr',
  lastModified: 2024-04-09T12:47:09.000Z,
  region: 'AMS',
  requestId: '871a9b69886a667c-AMS',
  signature: '8ae5f2c7f841088afa8962daaa6578427c5045fd6839d666cbb67ba0a6cebf7b'
}
```
After **altering the data in the database** and still within the 60 seconds of the SWR lifetime, the data is this:
```
{
  cacheStatus: 'swr',
  lastModified: 2024-04-09T12:47:09.000Z,
  region: 'AMS',
  requestId: '871a9ba8da170b8e-AMS',
  signature: '8ae5f2c7f841088afa8962daaa6578427c5045fd6839d666cbb67ba0a6cebf7b'
}
```
Subsequent requests all look like this but with a different requestId. No matter how many time I request it, the stale data from before altering of the database is being served. After the SWR lifetime has expired, the cacheStatus is 'miss' and the new data is being shown.

### Expected behavior (optional)

I expect the new data is shown the next time an SWR request serves the stale data and updates the data from the database. Is this a bug or am I missing something in the way SWR works?

### Information about Prisma Schema, Client Queries and Environment (optional)

- OS: Windows 11 WSL (Ubuntu)
- Database: MySQL 
- Node.js version: 21.7.1

- Run `prisma -v` to see your Prisma version and paste it
```
prisma                  : 5.11.0
@prisma/client          : 5.11.0
Computed binaryTarget   : debian-openssl-3.0.x
Operating System        : linux
Architecture            : x64
Node.js                 : v21.7.1
Query Engine (Node-API) : libquery-engine efd2449663b3d73d637ea1fd226bafbcf45b3102 (at node_modules/@prisma/engines/libquery_engine-debian-openssl-3.0.x.so.node)
Schema Engine           : schema-engine-cli efd2449663b3d73d637ea1fd226bafbcf45b3102 (at node_modules/@prisma/engines/schema-engine-debian-openssl-3.0.x)
Schema Wasm             : @prisma/prisma-schema-wasm 5.11.0-15.efd2449663b3d73d637ea1fd226bafbcf45b3102
Default Engines Hash    : efd2449663b3d73d637ea1fd226bafbcf45b3102
Studio                  : 0.499.0
```

**Comment by nurul3101** [comment]
Hey @JWW87 👋 

**During the TTL (Time to Live) Period**: The data is considered fresh, and any request within this period will return the data from the cache without hitting the database. In your case, this period is set to 15 seconds.

**After TTL Expires and Within SWR Time**: Once the TTL expires, the data enters the SWR phase, which lasts for 60 seconds in your setup. During this time, the cache will still serve the stale (now technically expired) data for any requests. Simultaneously, it will asynchronously refresh the data in the background. Once the data is refreshed in the cache, the updated data will be sent.

I hope this helps.

**Comment by JWW87** [comment]
Thanks for your answer @nurul3101! Looks like I don't fully understand the purpose of SWR. IMHO there's not much difference between TTL and SWR the way you describe it. I refactored my code to use the NextJS data caching layer instead of SWR as a workaround for my use case. Thanks again!

**Reply by pdparchitect** [reply]
I would like to second the comment above. SWR is practically useless under this condition.

So if SWR is set to 1 hour you will still push the old data but incur query costs?

I am sorry but this has zero benefits whatsoever and it is no different than setting the TTL with a minor exception that after expiry we will get fresh data. In fact, I would say it is kind of worse because processing happens regardless except that it never gets to see the light and day or be used in any shape or form.

Might as well not use accelerate and just cache the result with memcache.

**Reply by nurul3101** [reply]
@JWW87 @pdparchitect Thank you for bringing this to my attention. My understanding was incorrect. After clarifying with team internally, this is the correct behaviour:

> SWR mechanics

`swr` will not automatically refresh itself. Should an incoming request take place during this period, it will serve stale data and then go and fetch a refreshed value from the database.

Once the fresh data is retrieved and the cache is updated (even if still within the swr period), subsequent requests **will receive the fresh data**.

I'll use the below illustration to elaborate how ttl and swr works together;

It represents the below strategy;
```
cacheStrategy: {
  ttl: 60,
  swr: 30
}
```
Behaviour for an incoming request:

_0-60 seconds_
Data in the cache is considered fresh. Any incoming request will be served directly from the cache without contacting the database.

_60-90 seconds_
Once the TTL expires at 60 seconds, the data is considered stale. If a request comes in during this time 2 things happen;
stale data is served immediately to the client to minimize latency
in parallel, Accelerate makes a round trip and pulls latest data from the database and updates the cache.

_90+ seconds_
Considered a miss and will query the database directly.

We will update the docs to clarify this and will also update my comment above 🙏

**Reply by JWW87** [reply]
Thanks for your clarification @nurul3101. The way you describe it is exactly how I understood it from the docs. However, like I described in my first post, I don't see this behaviour when using the SWR strategy. 

This is what I see in practice (using your example): during the 60-90 seconds SWR period, stale data is still being served for every request, not just the first request after the TTL period. Even when a considerable amount of time has passed (long enough for Accelerate to pull new data from the database), the stale data is still served for the entire duration up until 90 seconds. Only when the SWR period has finished (90+ seconds), the fresh data will be presented.

So I agree with @pdparchitect that this is effectively the same as TTL. Are we missing something here, or is this a bug in Accelerate?

**Reply by nurul3101** [reply]
Do you mind reaching out via support@prisma.io email so that I can take a look at the logs for your project and debug what's going on?

**Reply by rajdtta** [reply]
@nurul3101 I'm still experiencing this behavior (SWR only returning stale data and not revalidating/refreshing it in the background). Was it not resolved/fixed?

**Reply by DavidKvas** [reply]
@nurul3101 I'm having precisely the same issue.

**Comment by pdparchitect** [comment]
Team I strongly believe you need reconsider this feature and re do it. At this point your caching strategy is not so useful for practical applications and I am sure most of your customers are suffering from this without even realising it.

"Stale While Refresh" means what it says, i.e. the data is stale but we are refreshing it and it will be fresh soon, perhaps within the next few moments.

It does not mean that you still get to see the old stale data until it expires. That makes no sense. This is effectively TTL.

I just did a couple of tests and with `{ swr: 3600 }` has the behaviour of ttl. Then what is the point?

**Reply by nurul3101** [reply]
@pdparchitect Apologies for the confusion, can you have a look at my comment [here](https://github.com/prisma/prisma/discussions/23787#discussioncomment-9237384)?
